funnel/models/email_address.py

```diff
@@ -0,0 +1,623 @@
+        """Record fact of a soft fail to this email address."""
+        self.delivery_state_at = db.func.utcnow()
+
+    @delivery_state.transition(None, delivery_state.HARD_FAIL)
+    def mark_hard_fail(self) -> None:
+        """Record fact of a hard fail to this email address."""
+        self.delivery_state_at = db.func.utcnow()
+
+    def refcount(self) -> int:
+        """Returns count of references to this EmailAddress instance"""
+        return sum(
+            len(getattr(self, backref_name)) for backref_name in self.__backrefs__
+        )
+
+    @classmethod
+    def mark_blocked(cls, email: str) -> None:
+        """
+        Mark email address as blocked.
+
+        Looks up all existing instances of EmailAddress with the same canonical
+        representation and amends them to forget the email address and set the
+        :attr:`is_blocked` flag.
+        """
+        for obj in cls.get_canonical(email, is_blocked=False).all():
+            obj.email = None
+            obj._is_blocked = True
+
+    @classmethod
+    def get_filter(
+        cls,
+        email: Optional[str] = None,
+        blake2b160: Optional[bytes] = None,
+        email_hash: Optional[str] = None,
+    ):
+        """
+        Get an filter condition for retriving an EmailAddress.
+
+        Accepts an email address or a blake2b160 hash in either bytes or base58 form.
+        Internally converts all lookups to a bytes-based hash lookup. Returns an
+        expression suitable for use as a query filter.
+        """
+        require_one_of(email=email, blake2b160=blake2b160, email_hash=email_hash)
+        if email:
+            blake2b160 = email_blake2b160_hash(email.lower())
+        elif email_hash:
+            blake2b160 = base58.b58decode(email_hash)
+
+        return cls.blake2b160 == blake2b160
+
+    @classmethod
+    def get(
+        cls,
+        email: Optional[str] = None,
+        blake2b160: Optional[bytes] = None,
+        email_hash: Optional[str] = None,
+    ) -> Optional[EmailAddress]:
+        """
+        Get an :class:`EmailAddress` instance by email address or its hash.
+
+        Internally converts an email-based lookup into a hash-based lookup.
+        """
+        return cls.query.filter(
+            cls.get_filter(email, blake2b160, email_hash)
+        ).one_or_none()
+
+    @classmethod
+    def get_canonical(
+        cls, email: str, is_blocked: Optional[bool] = None
+    ) -> Iterable[EmailAddress]:
+        """
+        Get :class:`EmailAddress` instances matching the canonical representation.
+
+        Optionally filtered by the :attr:`is_blocked` flag.
+        """
+        hashes = [
+            email_blake2b160_hash(result)
+            for result in canonical_email_representation(email.lower())
+        ]
+        query = cls.query.filter(cls.blake2b160_canonical.in_(hashes))
+        if is_blocked is not None:
+            query = query.filter_by(_is_blocked=is_blocked)
+        return query
```

Comment:
Should this be `.all()`? or returning query is fine?

Reply:
I'm moving away from enforcing `.all()` in `get*` methods.

1. The generator interface is a seamless replacement for a list
2. Additional filters can be applied if required

We are discouraging use of filters in views, but not in related models.